### PR TITLE
[WIP] remove translation of transport type for direct call object ID

### PR DESCRIPTION
**Pull request by zhijunfu**
## Why are these changes needed?

Currently we change the transport type of direct call object IDs from `DIRECT` to `RAYLET` when putting them to plasma store, this leads to extra complexity as well as cost (e.g. when deleting a direct call object we need to check plasma store as well). 

Ideally we can hold the invariant that there's always an entry in memory store (with `InPlasmaError`) for all the direct call object IDs in the plasma store. That would make things easier. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by ericl** [comment]
Can we hold off on these patches until we enable it by default? It is hard to tell if they impact correctness until we have all the tests enabled.

**Comment by ericl** [comment]
To summarize the new invariants:
- All remote call objects will be direct call type and present in either (1) memory store only, or (2) memory store and plasma (for promoted objects).
- Only objects placed in plasma with ray.put() will be plasma type.

Does this sounds right?

I wonder if we can also fix the ray.put case, to eliminate the idea of transport type entirely.

**Comment by zhijunfu** [comment]
> To summarize my understanding of new invariants proposed in this PR:
> 
> * All remote call objects will be direct call type and present in either (1) memory store only, or (2) memory store and plasma (for promoted objects).
> * Only objects placed in plasma with ray.put() will be plasma type.
> 
> Does this sounds right?

Yes that's right.

There might be some additional code changes needed, such as after pulling a direct call plasma object from remote, we also need to update the local memory store.

> 
> I wonder if we can also fix the ray.put case, to eliminate the idea of transport type entirely.

Good point. This might be doable. Probably good to try it out in a follow up PR.

**Comment by ericl** [comment]
Can you rebase this now to trigger a new test on master?

**Comment by zhijunfu** [comment]
> Can you rebase this now to trigger a new test on master?

done, thanks.

**Comment by ericl** [comment]
Looks like a lot of failing tests, a bunch might just be because of the check failure in plasma store that it shouldn't get direct ids.

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -492,12 +491,17 @@ Status CoreWorker::Contains(const ObjectID &object_id, bool *has_object) {
     // ErrorType::OBJECT_IN_PLASMA.
     found = memory_store_->Contains(object_id);
   }
+  /*
+  // We don't need this if we hold the invariant that all the direct call
```
Comment (edoakes) [inline src/ray/core_worker/core_worker.cc]: Not sure I understand this. According to the comment above, the memory store will return `false` if the object value is `OBJECT_IN_PLASMA`, so we still need to check, right?.

**Review by edoakes — COMMENTED** [review]
This looks great and will simplify the reference counting a lot. Let's try to get it merged soon . I removed the check and linted but couldn't push to your branch (didn't have permissions for some reason), so opened a new PR here for testing: https://github.com/ray-project/ray/pull/6531

**Comment by edoakes** [comment]
https://github.com/ray-project/ray/pull/6531 should be passing all tests now, so let's close this in favor of it.

**Comment by zhijunfu** [comment]
> #6531 should be passing all tests now, so let's close this in favor of it.

Sure, thanks

**Pull request closed without merging**